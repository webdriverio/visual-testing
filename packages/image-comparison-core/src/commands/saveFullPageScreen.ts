import beforeScreenshot from '../helpers/beforeScreenshot.js'
import afterScreenshot from '../helpers/afterScreenshot.js'
import { takeFullPageScreenshots } from '../methods/fullPageScreenshots.js'
import { makeFullPageBase64Image } from '../methods/images.js'
import type { ScreenshotOutput } from '../helpers/afterScreenshot.interfaces.js'
import type { BeforeScreenshotResult } from '../helpers/beforeScreenshot.interfaces.js'
import type { FullPageScreenshotDataOptions } from '../methods/screenshots.interfaces.js'
import type { InternalSaveFullPageMethodOptions } from './save.interfaces.js'
import { getMethodOrWicOption, canUseBidiScreenshot } from '../helpers/utils.js'
import { createBeforeScreenshotOptions, buildAfterScreenshotOptions } from '../helpers/options.js'

/**
 * Saves an image of the full page
 */
export default async function saveFullPageScreen(
    {
        browserInstance,
        instanceData,
        folders,
        tag,
        saveFullPageOptions,
        isNativeContext,
    }: InternalSaveFullPageMethodOptions
): Promise<ScreenshotOutput> {
    // 1a. Check if the method is supported in native context
    if (isNativeContext) {
        throw new Error('The method saveFullPageScreen is not supported in native context for native mobile apps!')
    }

    // 1b. Set some variables
    const { formatImageName, savePerInstance } = saveFullPageOptions.wic

    // 1c. Set the method options to the right values
    const enableLegacyScreenshotMethod = getMethodOrWicOption(saveFullPageOptions.method, saveFullPageOptions.wic, 'enableLegacyScreenshotMethod')
    const fullPageScrollTimeout = getMethodOrWicOption(saveFullPageOptions.method, saveFullPageOptions.wic, 'fullPageScrollTimeout')
    const hideAfterFirstScroll: HTMLElement[] = saveFullPageOptions.method.hideAfterFirstScroll || []
    const userBasedFullPageScreenshot = getMethodOrWicOption(saveFullPageOptions.method, saveFullPageOptions.wic, 'userBasedFullPageScreenshot')

    // 2.  Prepare the beforeScreenshot
    const beforeOptions = createBeforeScreenshotOptions(instanceData, saveFullPageOptions.method, saveFullPageOptions.wic)
    const enrichedInstanceData: BeforeScreenshotResult = await beforeScreenshot(browserInstance, beforeOptions, true)
    const {
        dimensions: {
            window: {
                devicePixelRatio,
                innerHeight,
                isEmulated: _isEmulated,
                isLandscape,
                screenHeight,
                screenWidth,
            },
        },
        isAndroid,
        isAndroidChromeDriverScreenshot,
        isAndroidNativeWebScreenshot,
        isIOS,
    } = enrichedInstanceData

    // 3.  Take fullpage screenshots with clean routing
    const fullPageScreenshotOptions: FullPageScreenshotDataOptions = {
        addressBarShadowPadding: beforeOptions.addressBarShadowPadding,
        devicePixelRatio: devicePixelRatio || NaN,
        deviceRectangles: instanceData.deviceRectangles,
        fullPageScrollTimeout,
        hideAfterFirstScroll,
        innerHeight: innerHeight || NaN,
        isAndroid,
        isAndroidChromeDriverScreenshot,
        isAndroidNativeWebScreenshot,
        isIOS,
        isLandscape,
        screenHeight: screenHeight || NaN,
        screenWidth: screenWidth || NaN,
        toolBarShadowPadding: beforeOptions.toolBarShadowPadding,
    }
    const shouldUseBidi = canUseBidiScreenshot(browserInstance) && (!userBasedFullPageScreenshot || !enableLegacyScreenshotMethod)
    const screenshotsData = await takeFullPageScreenshots(browserInstance, fullPageScreenshotOptions, shouldUseBidi)

    // 4.  Get the final image - either direct BiDi or stitched from multiple screenshots
    const fullPageBase64Image = (screenshotsData.fullPageHeight === -1 && screenshotsData.fullPageWidth === -1)
        ? screenshotsData.data[0].screenshot // BiDi screenshot - use directly
        : await makeFullPageBase64Image(screenshotsData, { devicePixelRatio: devicePixelRatio || NaN, isLandscape })

    // 5.  The after the screenshot methods
    const afterOptions = buildAfterScreenshotOptions({
        base64Image: fullPageBase64Image,
        folders,
        tag,
        isNativeContext: false,
        instanceData,
        enrichedInstanceData,
        beforeOptions,
        wicOptions: { formatImageName, savePerInstance }
    })

    // 6.  Return the data
    return afterScreenshot(browserInstance, afterOptions!)
}

