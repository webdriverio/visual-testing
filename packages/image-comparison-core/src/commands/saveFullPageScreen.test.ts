import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'
import saveFullPageScreen from './saveFullPageScreen.js'
import { createBeforeScreenshotOptions, buildAfterScreenshotOptions } from '../helpers/options.js'
import { takeFullPageScreenshots } from '../methods/takeFullPageScreenshots.js'
import { makeFullPageBase64Image } from '../methods/images.js'
import afterScreenshot from '../helpers/afterScreenshot.js'
import beforeScreenshot from '../helpers/beforeScreenshot.js'
import type { InternalSaveFullPageMethodOptions } from './save.interfaces.js'
import { BASE_CHECK_OPTIONS, createMethodOptions } from '../mocks/mocks.js'
import { canUseBidiScreenshot } from '../helpers/utils.js'

vi.mock('../helpers/beforeScreenshot.js', () => ({
    default: vi.fn().mockResolvedValue({
        browserName: 'chrome',
        browserVersion: '120.0.0',
        deviceName: 'desktop',
        dimensions: {
            window: {
                devicePixelRatio: 2,
                innerHeight: 900,
                isEmulated: false,
                isLandscape: false,
                outerHeight: 1000,
                outerWidth: 1200,
                screenHeight: 1080,
                screenWidth: 1920,
            },
        },
        isAndroid: false,
        isAndroidChromeDriverScreenshot: false,
        isAndroidNativeWebScreenshot: false,
        isIOS: false,
        isMobile: false,
        isTestInBrowser: true,
        logName: 'chrome',
        name: 'chrome',
        platformName: 'desktop',
        platformVersion: '120.0.0',
    })
}))
vi.mock('../methods/takeFullPageScreenshots.js', () => ({
    takeFullPageScreenshots: vi.fn().mockResolvedValue({
        fullPageHeight: 2000,
        fullPageWidth: 1200,
        data: [{
            canvasWidth: 1200,
            canvasYPosition: 0,
            imageHeight: 2000,
            imageWidth: 1200,
            imageXPosition: 0,
            imageYPosition: 0,
            screenshot: 'test-screenshot-data',
        }]
    })
}))
vi.mock('../methods/images.js', () => ({
    makeFullPageBase64Image: vi.fn().mockResolvedValue('fullpage-screenshot-data')
}))
vi.mock('../helpers/afterScreenshot.js', () => ({
    default: vi.fn().mockResolvedValue({
        devicePixelRatio: 2,
        fileName: 'test-fullpage.png'
    })
}))
vi.mock('../helpers/utils.js', () => ({
    canUseBidiScreenshot: vi.fn().mockReturnValue(false),
    getMethodOrWicOption: vi.fn().mockImplementation((method, wic, option) => {
        return method[option] ?? wic[option]
    })
}))
vi.mock('../helpers/options.js', () => ({
    createBeforeScreenshotOptions: vi.fn().mockReturnValue({
        instanceData: { test: 'data' },
        addressBarShadowPadding: 6,
        toolBarShadowPadding: 6,
        disableBlinkingCursor: false,
        disableCSSAnimation: false,
        enableLayoutTesting: false,
        hideElements: [],
        noScrollBars: false,
        removeElements: [],
        waitForFontsLoaded: false,
    }),
    buildAfterScreenshotOptions: vi.fn().mockReturnValue({
        actualFolder: '/test/actual',
        base64Image: 'fullpage-screenshot-data',
        disableBlinkingCursor: false,
        disableCSSAnimation: false,
        enableLayoutTesting: false,
        filePath: {
            browserName: 'chrome',
            deviceName: 'desktop',
            isMobile: false,
            savePerInstance: false,
        },
        fileName: {
            browserName: 'chrome',
            browserVersion: '120.0.0',
            deviceName: 'desktop',
            devicePixelRatio: 2,
            formatImageName: '{tag}-{browserName}-{width}x{height}',
            isMobile: false,
            isTestInBrowser: true,
            logName: 'chrome',
            name: 'chrome',
            outerHeight: 1000,
            outerWidth: 1200,
            platformName: 'desktop',
            platformVersion: '120.0.0',
            screenHeight: 1080,
            screenWidth: 1920,
            tag: 'test-fullpage'
        },
        hideElements: [],
        hideScrollBars: false,
        isLandscape: false,
        isNativeContext: false,
        platformName: 'desktop',
        removeElements: [],
    })
}))

describe('saveFullPageScreen', () => {
    let createBeforeScreenshotOptionsSpy: ReturnType<typeof vi.fn>
    let buildAfterScreenshotOptionsSpy: ReturnType<typeof vi.fn>
    let takeFullPageScreenshotsSpy: ReturnType<typeof vi.fn>
    let makeFullPageBase64ImageSpy: ReturnType<typeof vi.fn>
    let afterScreenshotSpy: ReturnType<typeof vi.fn>
    let canUseBidiScreenshotSpy: ReturnType<typeof vi.fn>
    let beforeScreenshotSpy: ReturnType<typeof vi.fn>

    const createBidiMockData = (screenshot: string) => ({
        fullPageHeight: -1,
        fullPageWidth: -1,
        data: [{
            canvasWidth: 0,
            canvasYPosition: 0,
            imageHeight: 0,
            imageWidth: 0,
            imageXPosition: 0,
            imageYPosition: 0,
            screenshot,
        }]
    })

    const baseOptions: InternalSaveFullPageMethodOptions = {
        browserInstance: { isAndroid: false, isMobile: false } as any,
        folders: BASE_CHECK_OPTIONS.folders,
        instanceData: BASE_CHECK_OPTIONS.instanceData,
        isNativeContext: false,
        saveFullPageOptions: {
            wic: BASE_CHECK_OPTIONS.wic,
            method: createMethodOptions({
                disableBlinkingCursor: false,
                disableCSSAnimation: false,
                enableLayoutTesting: false,
                enableLegacyScreenshotMethod: false,
                hideScrollBars: true,
                hideElements: [],
                removeElements: [],
                waitForFontsLoaded: true,
            })
        },
        tag: 'test-fullpage'
    }

    beforeEach(async () => {
        createBeforeScreenshotOptionsSpy = vi.mocked(createBeforeScreenshotOptions)
        buildAfterScreenshotOptionsSpy = vi.mocked(buildAfterScreenshotOptions)
        takeFullPageScreenshotsSpy = vi.mocked(takeFullPageScreenshots)
        makeFullPageBase64ImageSpy = vi.mocked(makeFullPageBase64Image)
        afterScreenshotSpy = vi.mocked(afterScreenshot)
        canUseBidiScreenshotSpy = vi.mocked(canUseBidiScreenshot)
        beforeScreenshotSpy = vi.mocked(beforeScreenshot)
    })

    afterEach(() => {
        vi.clearAllMocks()
    })

    it('should throw an error when in native context', async () => {
        const options = {
            ...baseOptions,
            isNativeContext: true
        }

        await expect(saveFullPageScreen(options)).rejects.toThrow(
            'The method saveFullPageScreen is not supported in native context for native mobile apps!'
        )
    })

    it('should use BiDi when conditions are met', async () => {
        canUseBidiScreenshotSpy.mockReturnValueOnce(true)
        takeFullPageScreenshotsSpy.mockResolvedValueOnce(createBidiMockData('test-bidi-screenshot-data'))

        const options = {
            ...baseOptions,
            saveFullPageOptions: {
                ...baseOptions.saveFullPageOptions,
                method: createMethodOptions({
                    userBasedFullPageScreenshot: false,
                    enableLegacyScreenshotMethod: false
                })
            }
        }

        await saveFullPageScreen(options)

        expect(takeFullPageScreenshotsSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should use BiDi when canUseBidiScreenshot=true, userBasedFullPageScreenshot=true, enableLegacyScreenshotMethod=false', async () => {
        canUseBidiScreenshotSpy.mockReturnValueOnce(true)
        takeFullPageScreenshotsSpy.mockResolvedValueOnce(createBidiMockData('test-bidi-screenshot-data-2'))

        const options = {
            ...baseOptions,
            saveFullPageOptions: {
                ...baseOptions.saveFullPageOptions,
                method: createMethodOptions({
                    userBasedFullPageScreenshot: true,
                    enableLegacyScreenshotMethod: false
                })
            }
        }

        await saveFullPageScreen(options)

        expect(takeFullPageScreenshotsSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should not use BiDi when canUseBidiScreenshot=true, userBasedFullPageScreenshot=false, enableLegacyScreenshotMethod=true', async () => {
        canUseBidiScreenshotSpy.mockReturnValueOnce(true)

        const options = {
            ...baseOptions,
            saveFullPageOptions: {
                ...baseOptions.saveFullPageOptions,
                method: createMethodOptions({
                    userBasedFullPageScreenshot: false,
                    enableLegacyScreenshotMethod: true
                })
            }
        }

        await saveFullPageScreen(options)

        expect(takeFullPageScreenshotsSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should take full page screenshots and return result', async () => {
        const result = await saveFullPageScreen(baseOptions)

        expect(createBeforeScreenshotOptionsSpy.mock.calls[0]).toMatchSnapshot()
        expect(takeFullPageScreenshotsSpy.mock.calls[0]).toMatchSnapshot()
        expect(buildAfterScreenshotOptionsSpy.mock.calls[0][0]).toMatchSnapshot()
        expect(makeFullPageBase64ImageSpy.mock.calls[0]).toMatchSnapshot()
        expect(afterScreenshotSpy.mock.calls[0][1]).toMatchSnapshot()
        expect(result).toEqual({
            devicePixelRatio: 2,
            fileName: 'test-fullpage.png'
        })
    })

    it('should handle missing dimension values with NaN fallbacks', async () => {
        beforeScreenshotSpy.mockResolvedValueOnce({
            browserName: 'chrome',
            browserVersion: '120.0.0',
            deviceName: 'desktop',
            dimensions: {
                body: {
                    scrollHeight: 2000,
                    offsetHeight: 1000,
                },
                html: {
                    clientWidth: 1200,
                    scrollWidth: 1200,
                    clientHeight: 1000,
                    scrollHeight: 2000,
                    offsetHeight: 1000,
                },
                window: {
                    devicePixelRatio: undefined,
                    innerHeight: undefined,
                    isEmulated: false,
                    isLandscape: false,
                    outerHeight: undefined,
                    outerWidth: undefined,
                    screenHeight: undefined,
                    screenWidth: undefined,
                },
            },
            isAndroid: false,
            isAndroidChromeDriverScreenshot: false,
            isAndroidNativeWebScreenshot: false,
            isIOS: false,
            isMobile: false,
            isTestInBrowser: true,
            logName: 'chrome',
            name: 'chrome',
            platformName: 'desktop',
            platformVersion: '120.0.0',
            isTestInMobileBrowser: false,
            addressBarShadowPadding: 6,
            toolBarShadowPadding: 6,
            appName: 'test-app',
            elementAddressBarPadding: 0,
            elementToolBarPadding: 0,
            pixelDensity: 1,
        } as any)

        const result = await saveFullPageScreen(baseOptions)

        expect(buildAfterScreenshotOptionsSpy.mock.calls[0][0]).toMatchSnapshot()
        expect(makeFullPageBase64ImageSpy.mock.calls[0]).toMatchSnapshot()
        expect(result).toEqual({
            devicePixelRatio: 2,
            fileName: 'test-fullpage.png'
        })
    })
})
