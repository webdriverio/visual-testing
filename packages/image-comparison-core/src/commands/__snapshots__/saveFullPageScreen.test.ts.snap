// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`saveFullPageScreen > should handle missing dimension values with NaN fallbacks 1`] = `
{
  "base64Image": "fullpage-screenshot-data",
  "beforeOptions": {
    "addressBarShadowPadding": 6,
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "hideElements": [],
    "instanceData": {
      "test": "data",
    },
    "noScrollBars": false,
    "removeElements": [],
    "toolBarShadowPadding": 6,
    "waitForFontsLoaded": false,
  },
  "enrichedInstanceData": {
    "addressBarShadowPadding": 6,
    "appName": "test-app",
    "browserName": "chrome",
    "browserVersion": "120.0.0",
    "deviceName": "desktop",
    "dimensions": {
      "body": {
        "offsetHeight": 1000,
        "scrollHeight": 2000,
      },
      "html": {
        "clientHeight": 1000,
        "clientWidth": 1200,
        "offsetHeight": 1000,
        "scrollHeight": 2000,
        "scrollWidth": 1200,
      },
      "window": {
        "devicePixelRatio": undefined,
        "innerHeight": undefined,
        "isEmulated": false,
        "isLandscape": false,
        "outerHeight": undefined,
        "outerWidth": undefined,
        "screenHeight": undefined,
        "screenWidth": undefined,
      },
    },
    "elementAddressBarPadding": 0,
    "elementToolBarPadding": 0,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isIOS": false,
    "isMobile": false,
    "isTestInBrowser": true,
    "isTestInMobileBrowser": false,
    "logName": "chrome",
    "name": "chrome",
    "pixelDensity": 1,
    "platformName": "desktop",
    "platformVersion": "120.0.0",
    "toolBarShadowPadding": 6,
  },
  "folders": {
    "actualFolder": "/test/actual",
    "baselineFolder": "/test/baseline",
    "diffFolder": "/test/diff",
  },
  "instanceData": {
    "appName": "TestApp",
    "browserName": "Chrome",
    "browserVersion": "118.0.0.0",
    "deviceName": "iPhone 14",
    "devicePixelRatio": 2,
    "deviceRectangles": {
      "bottomBar": {
        "height": 0,
        "width": 390,
        "x": 0,
        "y": 800,
      },
      "homeBar": {
        "height": 34,
        "width": 390,
        "x": 0,
        "y": 780,
      },
      "leftSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "rightSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "screenSize": {
        "height": 844,
        "width": 390,
      },
      "statusBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "statusBarAndAddressBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "viewport": {
        "height": 733,
        "width": 390,
        "x": 0,
        "y": 47,
      },
    },
    "initialDevicePixelRatio": 2,
    "isAndroid": false,
    "isIOS": true,
    "isMobile": true,
    "logName": "test-log",
    "name": "test-device",
    "nativeWebScreenshot": false,
    "platformName": "iOS",
    "platformVersion": "17.0",
  },
  "isNativeContext": false,
  "tag": "test-fullpage",
  "wicOptions": {
    "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
    "savePerInstance": true,
  },
}
`;

exports[`saveFullPageScreen > should handle missing dimension values with NaN fallbacks 2`] = `
[
  {
    "data": [
      {
        "canvasWidth": 1200,
        "canvasYPosition": 0,
        "imageHeight": 2000,
        "imageWidth": 1200,
        "imageXPosition": 0,
        "imageYPosition": 0,
        "screenshot": "test-screenshot-data",
      },
    ],
    "fullPageHeight": 2000,
    "fullPageWidth": 1200,
  },
  {
    "devicePixelRatio": NaN,
    "isLandscape": false,
  },
]
`;

exports[`saveFullPageScreen > should not use BiDi when canUseBidiScreenshot=true, userBasedFullPageScreenshot=false, enableLegacyScreenshotMethod=true 1`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addressBarShadowPadding": 6,
    "devicePixelRatio": 2,
    "deviceRectangles": {
      "bottomBar": {
        "height": 0,
        "width": 390,
        "x": 0,
        "y": 800,
      },
      "homeBar": {
        "height": 34,
        "width": 390,
        "x": 0,
        "y": 780,
      },
      "leftSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "rightSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "screenSize": {
        "height": 844,
        "width": 390,
      },
      "statusBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "statusBarAndAddressBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "viewport": {
        "height": 733,
        "width": 390,
        "x": 0,
        "y": 47,
      },
    },
    "fullPageScrollTimeout": 1500,
    "hideAfterFirstScroll": [],
    "innerHeight": 900,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isIOS": false,
    "isLandscape": false,
    "screenHeight": 1080,
    "screenWidth": 1920,
    "toolBarShadowPadding": 6,
  },
  true,
]
`;

exports[`saveFullPageScreen > should take full page screenshots and return result 1`] = `
[
  {
    "appName": "TestApp",
    "browserName": "Chrome",
    "browserVersion": "118.0.0.0",
    "deviceName": "iPhone 14",
    "devicePixelRatio": 2,
    "deviceRectangles": {
      "bottomBar": {
        "height": 0,
        "width": 390,
        "x": 0,
        "y": 800,
      },
      "homeBar": {
        "height": 34,
        "width": 390,
        "x": 0,
        "y": 780,
      },
      "leftSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "rightSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "screenSize": {
        "height": 844,
        "width": 390,
      },
      "statusBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "statusBarAndAddressBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "viewport": {
        "height": 733,
        "width": 390,
        "x": 0,
        "y": 47,
      },
    },
    "initialDevicePixelRatio": 2,
    "isAndroid": false,
    "isIOS": true,
    "isMobile": true,
    "logName": "test-log",
    "name": "test-device",
    "nativeWebScreenshot": false,
    "platformName": "iOS",
    "platformVersion": "17.0",
  },
  {
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "enableLegacyScreenshotMethod": false,
    "hideElements": [],
    "hideScrollBars": true,
    "removeElements": [],
    "waitForFontsLoaded": true,
  },
  {
    "addIOSBezelCorners": false,
    "addressBarShadowPadding": 6,
    "autoElementScroll": true,
    "autoSaveBaseline": false,
    "clearFolder": false,
    "compareOptions": {
      "blockOutSideBar": false,
      "blockOutStatusBar": false,
      "blockOutToolBar": false,
      "createJsonReportFiles": false,
      "diffPixelBoundingBoxProximity": 5,
      "ignoreAlpha": false,
      "ignoreAntialiasing": false,
      "ignoreColors": false,
      "ignoreLess": false,
      "ignoreNothing": false,
      "rawMisMatchPercentage": false,
      "returnAllCompareData": false,
      "saveAboveTolerance": 0,
      "scaleImagesToSameSize": false,
    },
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "enableLegacyScreenshotMethod": false,
    "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
    "fullPageScrollTimeout": 1500,
    "hideScrollBars": true,
    "isHybridApp": false,
    "savePerInstance": true,
    "tabbableOptions": {
      "circle": {
        "backgroundColor": "rgba(255, 0, 0, 0.4)",
        "borderColor": "rgba(255, 0, 0, 1)",
        "borderWidth": 1,
        "fontColor": "rgba(0, 0, 0, 1)",
        "fontFamily": "Arial",
        "fontSize": 10,
        "size": 10,
      },
      "line": {
        "color": "rgba(255, 0, 0, 1)",
        "width": 1,
      },
    },
    "toolBarShadowPadding": 6,
    "userBasedFullPageScreenshot": false,
    "waitForFontsLoaded": true,
  },
]
`;

exports[`saveFullPageScreen > should take full page screenshots and return result 2`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addressBarShadowPadding": 6,
    "devicePixelRatio": 2,
    "deviceRectangles": {
      "bottomBar": {
        "height": 0,
        "width": 390,
        "x": 0,
        "y": 800,
      },
      "homeBar": {
        "height": 34,
        "width": 390,
        "x": 0,
        "y": 780,
      },
      "leftSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "rightSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "screenSize": {
        "height": 844,
        "width": 390,
      },
      "statusBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "statusBarAndAddressBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "viewport": {
        "height": 733,
        "width": 390,
        "x": 0,
        "y": 47,
      },
    },
    "fullPageScrollTimeout": 1500,
    "hideAfterFirstScroll": [],
    "innerHeight": 900,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isIOS": false,
    "isLandscape": false,
    "screenHeight": 1080,
    "screenWidth": 1920,
    "toolBarShadowPadding": 6,
  },
  false,
]
`;

exports[`saveFullPageScreen > should take full page screenshots and return result 3`] = `
{
  "base64Image": "fullpage-screenshot-data",
  "beforeOptions": {
    "addressBarShadowPadding": 6,
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "hideElements": [],
    "instanceData": {
      "test": "data",
    },
    "noScrollBars": false,
    "removeElements": [],
    "toolBarShadowPadding": 6,
    "waitForFontsLoaded": false,
  },
  "enrichedInstanceData": {
    "browserName": "chrome",
    "browserVersion": "120.0.0",
    "deviceName": "desktop",
    "dimensions": {
      "window": {
        "devicePixelRatio": 2,
        "innerHeight": 900,
        "isEmulated": false,
        "isLandscape": false,
        "outerHeight": 1000,
        "outerWidth": 1200,
        "screenHeight": 1080,
        "screenWidth": 1920,
      },
    },
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isIOS": false,
    "isMobile": false,
    "isTestInBrowser": true,
    "logName": "chrome",
    "name": "chrome",
    "platformName": "desktop",
    "platformVersion": "120.0.0",
  },
  "folders": {
    "actualFolder": "/test/actual",
    "baselineFolder": "/test/baseline",
    "diffFolder": "/test/diff",
  },
  "instanceData": {
    "appName": "TestApp",
    "browserName": "Chrome",
    "browserVersion": "118.0.0.0",
    "deviceName": "iPhone 14",
    "devicePixelRatio": 2,
    "deviceRectangles": {
      "bottomBar": {
        "height": 0,
        "width": 390,
        "x": 0,
        "y": 800,
      },
      "homeBar": {
        "height": 34,
        "width": 390,
        "x": 0,
        "y": 780,
      },
      "leftSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "rightSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "screenSize": {
        "height": 844,
        "width": 390,
      },
      "statusBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "statusBarAndAddressBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "viewport": {
        "height": 733,
        "width": 390,
        "x": 0,
        "y": 47,
      },
    },
    "initialDevicePixelRatio": 2,
    "isAndroid": false,
    "isIOS": true,
    "isMobile": true,
    "logName": "test-log",
    "name": "test-device",
    "nativeWebScreenshot": false,
    "platformName": "iOS",
    "platformVersion": "17.0",
  },
  "isNativeContext": false,
  "tag": "test-fullpage",
  "wicOptions": {
    "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
    "savePerInstance": true,
  },
}
`;

exports[`saveFullPageScreen > should take full page screenshots and return result 4`] = `
[
  {
    "data": [
      {
        "canvasWidth": 1200,
        "canvasYPosition": 0,
        "imageHeight": 2000,
        "imageWidth": 1200,
        "imageXPosition": 0,
        "imageYPosition": 0,
        "screenshot": "test-screenshot-data",
      },
    ],
    "fullPageHeight": 2000,
    "fullPageWidth": 1200,
  },
  {
    "devicePixelRatio": 2,
    "isLandscape": false,
  },
]
`;

exports[`saveFullPageScreen > should take full page screenshots and return result 5`] = `
{
  "actualFolder": "/test/actual",
  "base64Image": "fullpage-screenshot-data",
  "disableBlinkingCursor": false,
  "disableCSSAnimation": false,
  "enableLayoutTesting": false,
  "fileName": {
    "browserName": "chrome",
    "browserVersion": "120.0.0",
    "deviceName": "desktop",
    "devicePixelRatio": 2,
    "formatImageName": "{tag}-{browserName}-{width}x{height}",
    "isMobile": false,
    "isTestInBrowser": true,
    "logName": "chrome",
    "name": "chrome",
    "outerHeight": 1000,
    "outerWidth": 1200,
    "platformName": "desktop",
    "platformVersion": "120.0.0",
    "screenHeight": 1080,
    "screenWidth": 1920,
    "tag": "test-fullpage",
  },
  "filePath": {
    "browserName": "chrome",
    "deviceName": "desktop",
    "isMobile": false,
    "savePerInstance": false,
  },
  "hideElements": [],
  "hideScrollBars": false,
  "isLandscape": false,
  "isNativeContext": false,
  "platformName": "desktop",
  "removeElements": [],
}
`;

exports[`saveFullPageScreen > should use BiDi when canUseBidiScreenshot=true, userBasedFullPageScreenshot=true, enableLegacyScreenshotMethod=false 1`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addressBarShadowPadding": 6,
    "devicePixelRatio": 2,
    "deviceRectangles": {
      "bottomBar": {
        "height": 0,
        "width": 390,
        "x": 0,
        "y": 800,
      },
      "homeBar": {
        "height": 34,
        "width": 390,
        "x": 0,
        "y": 780,
      },
      "leftSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "rightSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "screenSize": {
        "height": 844,
        "width": 390,
      },
      "statusBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "statusBarAndAddressBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "viewport": {
        "height": 733,
        "width": 390,
        "x": 0,
        "y": 47,
      },
    },
    "fullPageScrollTimeout": 1500,
    "hideAfterFirstScroll": [],
    "innerHeight": 900,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isIOS": false,
    "isLandscape": false,
    "screenHeight": 1080,
    "screenWidth": 1920,
    "toolBarShadowPadding": 6,
  },
  true,
]
`;

exports[`saveFullPageScreen > should use BiDi when conditions are met 1`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addressBarShadowPadding": 6,
    "devicePixelRatio": 2,
    "deviceRectangles": {
      "bottomBar": {
        "height": 0,
        "width": 390,
        "x": 0,
        "y": 800,
      },
      "homeBar": {
        "height": 34,
        "width": 390,
        "x": 0,
        "y": 780,
      },
      "leftSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "rightSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "screenSize": {
        "height": 844,
        "width": 390,
      },
      "statusBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "statusBarAndAddressBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "viewport": {
        "height": 733,
        "width": 390,
        "x": 0,
        "y": 47,
      },
    },
    "fullPageScrollTimeout": 1500,
    "hideAfterFirstScroll": [],
    "innerHeight": 900,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isIOS": false,
    "isLandscape": false,
    "screenHeight": 1080,
    "screenWidth": 1920,
    "toolBarShadowPadding": 6,
  },
  true,
]
`;
